secboot: re-introduce v1 fde hook format

valentindavid opened the pull request.

bboozzoo commented on secboot/secboot_hooks.go:
```diff
@@ -118,21 +122,45 @@ func SealKeysWithFDESetupHook(runHook fde.RunSetupHookFunc, keys []SealKeyReques
-	if _, err := io.ReadFull(f, buf); err != nil {
-		return false
+	params := sb.KeyParams{
+		Handle:       json.RawMessage(handle),
```
can you leave a comment stating that we intentionally store key data in the handle as it is treated as a opaque blob by secboot?

bboozzoo commented on secboot/secboot_hooks.go:
```diff
@@ -118,21 +122,45 @@ func SealKeysWithFDESetupHook(runHook fde.RunSetupHookFunc, keys []SealKeyReques
+
+func (fh *fdeHookV1DataHandler) RecoverKeys(data *sb.PlatformKeyData, encryptedPayload []byte) ([]byte, error) {
+	var handle []byte
+	if err := json.Unmarshal(data.EncodedHandle, &handle); err != nil {
```
could use a comment
```suggestion
	// we abuse the handle to store the actual blob carrying sealed key
	if err := json.Unmarshal(data.EncodedHandle, &handle); err != nil {
```

valentindavid replied:
I removed this code

bboozzoo reviewed (approved):
LGTM

valentindavid commented:
This was manually tested with
* snapd 2.49.2
* removing the assume in the pc gadget from latest/edge
* rebuild pc-kernel latest/edge with core-initrd core20 branch and snap-bootstrap from this PR
* core20 from latest/edge

bboozzoo commented on secboot/secboot_sb.go:
```diff
@@ -161,6 +161,22 @@ func UnlockVolumeUsingSealedKeyIfEncrypted(disk disks.Disk, name string, sealedE
 		return res, fmt.Errorf("internal error: cannot build an auth requestor: %v", err)
 	}
 
+	if loadedKey.SealedKeyV1 != nil {
```
this feels like it's a leaky abstraction which only muddies the whole picture, could we do something more explicit, eg:

```
// legacy fde-hooks keys which did not carry any metadata
var legacyV1Key []byte
// more modern keys with metadata
var keys []*sb.KeyData

// isLegacyV1Key - simply checks the header

if fdeHasRevealKey() && isLegacyV1Key(sealedEncryptioKeyFile) {
	legacyV1Key, err = os.ReadFile()
	...
} else {
	if err := readKeyFile(sealedEncryptionKeyFile, loadedKey); err != nil {
		return res, err
	}
}

...
if len(legacyV1KeyFile) != 0 {

}
...
// continue with ActivateVolumeWithKeyData
```

and then simply drop the loader thingy

bboozzoo reviewed (approved):
LGTM

(although I added some changes here, so it should likely count as +0.5)

valentindavid commented:
I have retested. And it is broken.

valentindavid commented:
It is working again.

bboozzoo commented on secboot/secboot_tpm.go:
```diff
@@ -238,14 +238,34 @@ func newAuthRequestor() (sb.AuthRequestor, error) {
-// TODO: consider moving this to secboot
-func readKeyFileImpl(keyfile string) (*sb.KeyData, *sb_tpm2.SealedKeyObject, error) {
+type keyLoader interface {
+	LoadKeyData(kd *sb.KeyData)
+	LoadSealedKeyObject(sko *sb_tpm2.SealedKeyObject)
+	LoadFDEHookKeyV1(sk []byte)
```
```suggestion
	// LoadKeyData is expected to keep track of key data wrappers
	LoadKeyData(kd *sb.KeyData)
	// LoadSealedKeyObject is expected to keep track of a TPM sealed key
	LoadSealedKeyObject(sko *sb_tpm2.SealedKeyObject)
	// LoadFDEHookV1Key is expected to keep track of a FDE hook v1 sealed key
	LoadFDEHookKeyV1(sk []byte)
```

bboozzoo commented on secboot/secboot_tpm.go:
```diff
@@ -254,25 +274,58 @@ func readKeyFileImpl(keyfile string) (*sb.KeyData, *sb_tpm2.SealedKeyObject, err
+	case oldSealedKey && hintExpectFDEHook:
+		// FDE hook key v1
+		//
+		// It has the same magic header has sealed key object,
```
```suggestion
		// It has the same magic header as sealed key object,
```

bboozzoo commented on secboot/secboot_tpm.go:
```diff
@@ -254,25 +274,58 @@ func readKeyFileImpl(keyfile string) (*sb.KeyData, *sb_tpm2.SealedKeyObject, err
+	return l == len(rawPrefix) && bytes.HasPrefix(buf, rawPrefix), nil
+}
+
+// TODO: consider moving this to secboot
```
```suggestion
// readKeyFile attempts to read the provided key file. Will call the key loader's methods
// appropriate for a given key artifact passing key type specific data structure carrying
// the relevant data.
// TODO: consider moving this to secboot
```

bboozzoo reviewed (approved):
LGTM, just some minor comments

ZeyadYasser reviewed (commented):
Small comments

valentindavid commented:
I had to resolve conflicts, so I squashed and and rebase. I also added some code in the resealing to ignore v1 keys and a test for it. So that will need to be reviewed.

valentindavid commented on secboot/secboot_tpm.go:
```diff
@@ -238,14 +238,40 @@ func newAuthRequestor() (sb.AuthRequestor, error) {
 
-// TODO: consider moving this to secboot
-func readKeyFileImpl(keyfile string) (*sb.KeyData, *sb_tpm2.SealedKeyObject, error) {
+type keyLoader interface {
```
I had in mind that in the future we just make callbacks that ignore some input. For instance the sealed key objects for tpm are useless for unlocking, only for resealing. Though in the end I have just used the same type.

valentindavid replied:
I will try it and see if the code looks better. Maybe it can be done in another PR.

pedronis replied:
Wonder if the interface should be called loadedKey now?

pedronis replied:
also are we going to have other implementations of this interface in later PRs?

valentindavid replied:
Yes. But I am not sure how it should look at this point. But one thing is sure is that having an interface here seemed a bit useless.

pedronis commented on secboot/secboot_hooks.go:
```diff
@@ -160,14 +168,26 @@ func ResealKeysWithFDESetupHook(keyFiles []string, primaryKeyFile string, models
 	return nil
 }
 
+func unlockDiskWithHookV1Key(mapperName, sourceDevice string, sealed []byte) error {
```
tests don't reach here

valentindavid replied:
The tests for it were botched. Also I have noticed that TestUnlockVolumeUsingSealedKeyIfEncryptedFdeRevealKeyV1AndV1GeneratedSealedKeyFile and TestUnlockVolumeUsingSealedKeyIfEncryptedFdeRevealKeyV1 were almost the same, just some checks were different. So I also kept only one with all the checks.

pedronis commented on secboot/secboot_sb.go:
```diff
@@ -162,6 +164,23 @@ func UnlockVolumeUsingSealedKeyIfEncrypted(disk disks.Disk, name string, sealedE
+	if loadedKey.FDEHookKeyV1 != nil {
+		// Special case for hook keys v1. They do not have
+		// primary keys. So we cannot wrap them in KeyData
+		err := unlockDiskWithHookV1Key(mapperName, sourceDevice, loadedKey.FDEHookKeyV1)
```
they also don't reach here

pedronis commented on secboot/secboot_tpm.go:
```diff
@@ -238,14 +238,40 @@ func newAuthRequestor() (sb.AuthRequestor, error) {
-func readKeyFileImpl(keyfile string) (*sb.KeyData, *sb_tpm2.SealedKeyObject, error) {
+type keyLoader interface {
+	// LoadKeyData keeps track of keys in KeyData format.
+	LoadKeyData(kd *sb.KeyData)
```
wouldn't Loaded... be clearer?  this just keep tracks at the comment says, the loading has already happened

pedronis reviewed (commented):
some comments

pedronis commented on secboot/secboot_sb.go:
```diff
@@ -162,6 +164,23 @@ func UnlockVolumeUsingSealedKeyIfEncrypted(disk disks.Disk, name string, sealedE
+		// with key data if there are some on the tokens.
+		// Also the request for recovery key will happen in
+		// ActivateVolumeWithKeyData
+		logger.Noticef("WARNING: attempting opening device %s  with key file %s failed: %v", sourceDevice, sealedEncryptionKeyFile, err)
```
this fallback path is not reached

valentindavid replied:
I have added `TestUnlockVolumeUsingSealedKeyIfEncryptedFdeRevealKeyV1FallbackKeyData

pedronis commented on secboot/secboot_tpm.go:
```diff
@@ -238,14 +238,40 @@ func newAuthRequestor() (sb.AuthRequestor, error) {
+	LoadedFDEHookKeyV1(sk []byte)
+}
+
+type defaultKeyLoader struct {
```
then this is also defaultLoadedKey?

valentindavid replied:
I have put a todo to the `keyLoader` interface to change it.

I am not sure that `defaultLoadedKey` makes total sense here since we create it empty. So it is not loaded yet. But it will make sense once we get rid of the interface and we return it.

pedronis commented on secboot/secboot_tpm.go:
```diff
@@ -433,30 +500,33 @@ func ResealKeys(params *ResealKeysParams) error {
-			hasNewData = true
-		} else {
-			hasOldObject = true
+		if loadedKey.SealedKeyObject != nil {
```
we should maybe leave a TODO about refactor some of these  ifs into methods on the loadedKey interface, but also I as wrote elsewhere wondering if there will be other implementations?

valentindavid replied:
I have added a TODO to the definition of `keyLoader` interface.

pedronis reviewed (commented):
thanks, couple naming comments and some more questions/comments

pedronis reviewed (approved):
thanks

ZeyadYasser reviewed (approved):
LGTM, Thank you!

This could use a spread test in a follow up PR.

bboozzoo reviewed (approved):
Thanks. The cleanups can be done in a followup PR.

chrisccoulson reviewed (approved):
This looks ok to me, although I've not spent a lot of time looking at the test code.
